wled00/bus_manager.h

```diff
@@ -21,89 +22,44 @@ uint16_t approximateKelvinFromRGB(uint32_t rgb);
 
-//temporary struct for passing bus configuration to bus
-struct BusConfig {
-  uint8_t type;
-  uint16_t count;
-  uint16_t start;
-  uint8_t colorOrder;
-  bool reversed;
-  uint8_t skipAmount;
-  bool refreshReq;
-  uint8_t autoWhite;
-  uint8_t pins[5] = {255, 255, 255, 255, 255};
-  uint16_t frequency;
-  bool doubleBuffer;
-  uint8_t milliAmpsPerLed;
-  uint16_t milliAmpsMax;
-
-  BusConfig(uint8_t busType, uint8_t* ppins, uint16_t pstart, uint16_t len = 1, uint8_t pcolorOrder = COL_ORDER_GRB, bool rev = false, uint8_t skip = 0, byte aw=RGBW_MODE_MANUAL_ONLY, uint16_t clock_kHz=0U, bool dblBfr=false, uint8_t maPerLed=LED_MILLIAMPS_DEFAULT, uint16_t maMax=ABL_MILLIAMPS_DEFAULT)
-  : count(len)
-  , start(pstart)
-  , colorOrder(pcolorOrder)
-  , reversed(rev)
-  , skipAmount(skip)
-  , autoWhite(aw)
-  , frequency(clock_kHz)
-  , doubleBuffer(dblBfr)
-  , milliAmpsPerLed(maPerLed)
-  , milliAmpsMax(maMax)
-  {
-    refreshReq = (bool) GET_BIT(busType,7);
-    type = busType & 0x7F;  // bit 7 may be/is hacked to include refresh info (1=refresh in off state, 0=no refresh)
-    size_t nPins = 1;
-    if (IS_VIRTUAL(type))   nPins = 4; //virtual network bus. 4 "pins" store IP address
-    else if (IS_2PIN(type)) nPins = 2;
-    else if (IS_PWM(type))  nPins = NUM_PWM_PINS(type);
-    for (size_t i = 0; i < nPins; i++) pins[i] = ppins[i];
-  }
-
-  //validates start and length and extends total if needed
-  bool adjustBounds(uint16_t& total) {
-    if (!count) count = 1;
-    if (count > MAX_LEDS_PER_BUS) count = MAX_LEDS_PER_BUS;
-    if (start >= MAX_LEDS) return false;
-    //limit length of strip if it would exceed total permissible LEDs
-    if (start + count > MAX_LEDS) count = MAX_LEDS - start;
-    //extend total count accordingly
-    if (start + count > total) total = start + count;
-    return true;
-  }
-};
-
+struct BusConfig; // forward declaration
 
 // Defines an LED Strip and its color ordering.
-struct ColorOrderMapEntry {
+typedef struct {
   uint16_t start;
   uint16_t len;
   uint8_t colorOrder;
-};
+} ColorOrderMapEntry;
 
 struct ColorOrderMap {
-    void add(uint16_t start, uint16_t len, uint8_t colorOrder);
+    bool add(uint16_t start, uint16_t len, uint8_t colorOrder);
 
-    uint8_t count() const { return _count; }
+    inline uint8_t count() const { return _mappings.size(); }
 
     void reset() {
-      _count = 0;
-      memset(_mappings, 0, sizeof(_mappings));
+      _mappings.clear();
+      _mappings.shrink_to_fit();
     }
 
     const ColorOrderMapEntry* get(uint8_t n) const {
-      if (n > _count) {
-        return nullptr;
-      }
+      if (n >= count()) return nullptr;
       return &(_mappings[n]);
     }
 
     uint8_t getPixelColorOrder(uint16_t pix, uint8_t defaultColorOrder) const;
 
   private:
-    uint8_t _count;
-    ColorOrderMapEntry _mappings[WLED_MAX_COLOR_ORDER_MAPPINGS];
+    std::vector<ColorOrderMapEntry> _mappings;
```

Comment:
Is the switch to `std::vector` intended to reduce memory usage?

One of the challenges is that `std::vector` has an exponential allocation strategy.  Depending on the relationship between WLED_MAX_COLOR_ORDER_MAPPINGS and the number of entries actually used, it may end up using *more* memory.  I checked the code for std::vector for ESP32 -- this implementation of `push_back` uses the classic "double the array length if it needs more memory" strategy.  So for `WLED_MAX_COLOR_ORDER_MAPPINGS == 10` (esp32), when you push the 9th element on to the list, the buffer will expand to 16 -- leaving 6 elements you pay for in RAM, but will never use.

What you can do to mitigate this is any of:
- You could ensure that the limits are powers of 2
- When loading the vector, figure out how many elements you need, then call `reserve()` up front to allocate exactly the right amount of memory
- Alternately, you can preallocate the maximum memory you plan to use with `reserve()` at construct time; won't save any RAM but also guarantees you'll never allocate more than you need.
- You could call `shrink_to_fit()` after filling the vector, or make a copy and discard the original std::vector object.  (Cost is the same, a complete copy and temporarily holding more RAM than it needs)

Reply:
> Is the switch to `std::vector` intended to reduce memory usage?

No. It is to allow variable size (in the future) without the expense of pre-allocated memory. The question was "why 10 and why only 10?".
`reserve()` could be done, yes. That'll go into `cfg.cpp` and `set.cpp`. Thanks.